### PR TITLE
Fix #8108, c01a2e2: possible crash on loading TTD savegames with phantom oil rigs

### LINKED ISSUE #8108: Loading old TTD save crashes game (station / oil rig related)
## Version of OpenTTD
Release: 1.10.1
Nightly: 20200426

## Expected result
In both versions, the savegame loads correctly.

## Actual result (Release):
The game loads normally, but crashes after a few days if you demolish the bridge located at tile 0x299A (154, 41).
Crash log (after demolishing everything): [crash_oil_rig_1.zip](https://github.com/OpenTTD/OpenTTD/files/4573793/crash_oil_rig_1.zip)
Crash log (after demolishing bridge at 0x299A): [crash_floating_train.zip](https://github.com/OpenTTD/OpenTTD/files/4573929/crash_floating_train.zip)

Additionally, demolishing the bridge at tile 0x4181 (129, 65) causes the train on it to perpetually say "No Power".

## Actual result (Nightly): 
The game asserts here:
https://github.com/OpenTTD/OpenTTD/blob/fde28ec0b0b85451bfa80345d2b45e92ddba1655/src/station_cmd.cpp#L4205
This is called during the station-based tileloop at https://github.com/OpenTTD/OpenTTD/blob/57553cd8098ac003b3d5df6b9778214e6933c5d7/src/saveload/afterload.cpp#L945
Crash log: [crash_oil_rig_nightly.zip](https://github.com/OpenTTD/OpenTTD/files/4573795/crash_oil_rig_nightly.zip)

## Steps to Reproduce
1. In a release version, load the attached savegame.
2. Pause the game.
3. Give yourself a few billion.
4. Demolish everything.
5. Wait up to 30 days.
6. Crash

1. In a nightly version, load the attached savegame.
2. Crash

## Savegame
[#3601 - TRT01.zip](https://github.com/OpenTTD/OpenTTD/files/4573788/3601.-.TRT01.zip)

### PR DESCRIPTION
**Pull request by SamuXarick**
Need help with commit message. Also, I'm unsure if the iterator issue originated from ab711e6. Plz someone check.

### CODE REVIEW

**Comment by Yexo** [comment]
[c01a2e](https://github.com/OpenTTD/OpenTTD/commit/c01a2e2a81d8e7bcd47d46292ed0b7d452081c31) is the actual commit that swapped out SmallVector with std::vector, and changed the semantics of the erase function which causes this crash.

File: src/saveload/afterload.cpp [hunk 1]
```diff
@@ -929,19 +929,18 @@ bool AfterLoadGame()
 						break;
 
 					case STATION_OILRIG: {
+						/* The internal encoding of oil rigs was changed twice.
+						 * It was 3 (till 2.2) and later 5 (till 5.1).
+						 * Setting it unconditionally does not hurt.
+						 */
+						Station::GetByTile(t)->airport.type = AT_OILRIG;
```
Comment (Yexo) [inline src/saveload/afterload.cpp:937]: Can you modify the last sentence of the comment to something like this which explains why it's moved out of the if/else block like before:
DeleteOilRig asserts on the correct type, and setting it unconditionally does not hurt.

File: src/saveload/afterload.cpp [hunk 2]
```diff
@@ -2203,7 +2202,8 @@ bool AfterLoadGame()
 			}
 
 			if (remove) {
-				DeleteAnimatedTile(*tile);
+				tile = _animated_tiles.erase(tile);
+				MarkTileDirtyByTile(*tile);
```
Comment (ldpl) [inline src/saveload/afterload.cpp]: Why do you even need to mark dirty tiles in afterlolad?

Comment (SamuXarick) [reply src/saveload/afterload.cpp]: I copied over from DeleteAnimatedTile. But if it doesn't need that, I can remove this line.

**Comment by Yexo** [comment]
Suggested commit message: Fix #8108, c01a2e2: possible crash on loading TTD savegames with phantom oil rigs

**Review by Yexo — APPROVED** [review]

**Pull request merged**